### ISSUE TITLE
getIdAndSecret header vs body credentials validation order

### ISSUE THREAD

**Issue by jcferrer** [apigee-127/volos#95]
Scott,

I found this edge case. Let me give some background first (kind of long, but please read).

I am using a127-magic with volos-oauth-redis and swagger-tools-embedded swagger-ui. I have an oauth2 app, running on this framework that basically exposes the basic oauth operations (authorize, access_token, refresh, invalidate). 
The POST /access_token takes the expected body format:
````
{
  "grant_type": "client_credentials",
  "client_id": "myId",
  "client_secret": "mySecret"
}
````
All good so far. No issues.

To protect this end point from unauthorized external access, I decided to set Basic Authentication challenge using Nginx in front of my app.

This is where I found a problem. After passing the Swagger-UI Basic Auth challenge, and trying to POST a client request body to generate a token, I found the UI kept popping up the credentials window, even though I entered them while accessing the UI the first time.

After spending a lot of time looking everywhere without success, I then found that the function ``` getIdAndSecret``` located the ```oauth.js``` file, favors the presence of the Authorization Header for Basic Auth credentials, over the credentials provided in the request body. In my case, this is an issue, since the Authorization Header is already carrying the Basic Auth credentials entered when first landing on the UI, so the ones I provide in the request body are not taken, resulting in an invalid_client error.

Here's the function:

```
function getIdAndSecret(authorizeHeader, parsedBody) {
  // 2.3.1: Client id and secret may be in Basic format, or in the request body
  var clientId;
  var clientSecret;
  if (authorizeHeader) {
    var parsedHeader = /Basic (.+)/.exec(authorizeHeader);
    if (!parsedHeader) {
      return null;
    }
    if (parsedHeader.length < 2) {
      return null;
    }

    var decodedHeader = new Buffer(parsedHeader[1], 'base64').toString();
    var decoded = /([^:]+):([^:]+)/.exec(decodedHeader);
    if (!decoded || decoded.length < 3) {
      return null;
    }

    clientId = decoded[1];
    clientSecret = decoded[2];
  } else {
    clientId = parsedBody.client_id;
    clientSecret = parsedBody.client_secret;
  }
  if (clientId && clientSecret) {
    return [clientId, clientSecret];
  }
  return null;
}
```

For now to solve the problem I fixed the condition to be:

```
 if (authorizeHeader && !parsedBody.client_id && !parsedBody.client_secret) {
```

in order to force using the body credentials if present. I am sure you will find a better way to do it though. 

Do you think you can reverse the order of checking, favoring the body credentials if present?

Let me know what you think.

Thanks

JC

**Comment by theganyo** [comment]
Hmm. I guess I don't see a problem with this. I find it hard to imagine a scenario where one would include the client_id and client_secret in the body yet still intend to use the authorization header and I can imagine others might run into a similar situation as yourself.

**Comment by jcferrer** [comment]
I guess the whole thing about protecting the actual Swagger UI endpoints from unauthorized access makes it very tricky. So far the change I made is working for me. I don't like it because I had to fork.

My other option was to not expose my access_token operation with a swagger UI, but I did it in the spirit of illustrating the token generation process. I still find a lot of developers struggling with this in the beginning.

Anyways this whole issue is just because I decided to use Swagger UI to expose the OAuth2 token operations and protect its UI with Basic Auth.

Let me know if you feel the change I proposed is worth trying.

Thanks for your help.
